doc/rst/language/spec/procedures.rst

```diff
@@ -1071,8 +1071,9 @@ value of the invoking call expression.
 A return statement in a procedure of a non-\ ``void`` return type
 (:ref:`Return_Types`) must include an expression. A return
 statement in a procedure of a ``void`` return type or in an iterator
-must not include an expression. A return statement of a variable
-procedure must contain an lvalue expression.
+must not include an expression. A return statement of a procedure with
+``ref`` or ``const ref`` return intent must contain an lvalue
+expression.
```

Comment:
Thanks for following through with this, Paul!  Though you inherited this, I think "of a procedure" should be "in a procedure" to match the previous sentence.  And if you fix that, the word "contain" is tripping me up as well... though it's hard to come up with one-word alternatives.  Maybe "involve"?  "refer to"?

Reply:
How about "the expression included in return statements  in procedures with ``ref`` or ``const ref`` return intent must be an lvalue expression."?

More drastically, I might move the void return type case to be first or last, not between the two cases that deal with non-void returns.  That would let the ref sentence be just "Further, if the procedure has ``ref`` or ``const ref`` return intent, that expression must be an lvalue expression", following on directly from the now-previous sentence.


("of a void return type", or "with a void return type"?)

Reply:
Since we're considering more dramatic rewrites, if we were sticking with the current flow, what do you think about "For procedures with ``ref`` or ``const ref`` return intents, the expression being returned must be an lvalue expression?" or even "...must be an expression with storage associated with it that will outlive the procedure itself"?  (since 'lvalue' is a pretty jargon-y term).    To me, part of the problem with the current wording is that it feels a bit like it's trying to jump through hoops to refer to "that expression that is part of a return statement, you know the one we're talking about, right?" when it seems OK to assume that readers will understand what we're talking about at this point without referring to "return statement" and the expression that is in it / contained by it / other awkward preposition here rather than just referring to the returned expression?  

> More drastically, I might move the void return type case to be first or last, not between the two cases that deal with non-void returns. That would let the ref sentence be just "Further, if the procedure has ref or const ref return intent, that expression must be an lvalue expression", following on directly from the now-previous sentence.

That makes sense to me as well.  Following on with my previous comment I might say one of "the [returned] expression..." to avoid the slightly awkward "that expression".  And I might put in the longer (and maybe more accurate) replacement for lvalue expression if you like that.  (specifically, I think a local variable of type int has an lvalue, but shouldn't be returned by a ref/const ref intent).

Reply:
I made those changes, though I wonder

* if the language specification is technical enough that it could assume the reader is familiar with lvalues.
  * if not, if it could define the term, since it seems like a useful concept.
* if the storage that will outlive the procedure call is a concept the spec gives a name to elsewhere (if so, I couldn't find it.)
  * if not, if it could be.  (A sentence about one of the cases in "The Return Statement" seems a kind of awkward place to introduce the idea.)
* if the lifetime of the expression is a concern to the return statement itself, or if there's a more general lifetime section.  On a related note, the classes spec page says accessing a class variable after deleting it results in undefined behavior, not that it must not be done.

Reply:
Thanks for the edits Paul!  I'm inclined to merge this, but will wait to hear from you before doing so:

I think you're right that it's reasonable for the language spec to assume the reader is familiar with lvalues, or to define the term then use it—and with a quick check, I see we use it in a number of places already.  I think I was getting mixed up with error messages, where we've tried to avoid the term.  That said, I think the additional notion that it continue to be a legal lvalue after the procedure returns is important, which was my secondary motivation for that rewording.

I agree that this is a concept that might be worth promoting to a well-defined term of its own in the spec.  I think the edit here is worthwhile on its own merits without needing to wrestle that bigger question to the ground, though (which is not to say that you couldn't/shouldn't wrestle with it further if you're motivated to do so—I just wouldn't want to have this PR get stuck behind that effort given that I consider it a net positive).

Reply:
Fine by me to merge.  I'm not planning to wrestle with the spec at large.  :)

Reply:
Nor was I expecting you to.  I appreciate you taking the time to fix this, and then wrestle with me on the wording, having noticed the issue!